lib/internal/worker/messaging.js

```diff
@@ -0,0 +1,229 @@
+    AtomicsNotify(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+    return;
+  }
+
+  port.postMessage(
+    {
+      type: messageTypes.RECEIVE_MESSAGE_FROM_WORKER,
+      source,
+      destination,
+      value,
+      memory,
+    },
+    transferList,
+  );
+}
+
+function receiveMessageFromWorker(source, value, memory) {
+  const status = new Int32Array(memory);
+  const deliverable = Number(process.listenerCount('workerMessage') > 0);
+
+  // If somebody is listening, emit the message
+  if (deliverable) {
+    process.emit('workerMessage', value, source);
+  }
+
+  // Populate the result
+  AtomicsStore(status, WORKER_MESSAGING_RESULT_INDEX, deliverable);
+  AtomicsStore(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+  AtomicsNotify(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+}
+
+function createMainThreadPort(thread) {
+  // Create a channel that links the new thread to the main thread
+  const {
+    port1: mainThreadPortToMain,
+    port2: mainThreadPortToThread,
+  } = new MessageChannel();
+
+  const registrationMessage = {
+    type: messageTypes.REGISTER_MAIN_THREAD_PORT,
+    thread,
+    port: mainThreadPortToMain,
+  };
+
+  if (isMainThread) {
+    handleMessageFromThread(registrationMessage);
+  } else {
+    mainThreadPort.postMessage(registrationMessage, [mainThreadPortToMain]);
+  }
+
+  return mainThreadPortToThread;
+}
+
+function destroyMainThreadPort(thread) {
+  const unregistrationMessage = {
+    type: messageTypes.UNREGISTER_MAIN_THREAD_PORT,
+    thread,
+  };
+
+  if (isMainThread) {
+    handleMessageFromThread(unregistrationMessage);
+  } else {
+    mainThreadPort.postMessage(unregistrationMessage);
+  }
+}
+
+function setupMainThreadPort(port) {
+  mainThreadPort = port;
+  mainThreadPort.on('message', handleMessageFromMainThread);
+
+  // Never block the process on this port
+  mainThreadPort.unref();
+}
+
+async function postMessageToWorker(destination, value, transferList, timeout) {
+  if (typeof transferList === 'number' && typeof timeout === 'undefined') {
+    timeout = transferList;
+    transferList = [];
+  }
+
+  if (typeof timeout !== 'undefined') {
+    validateNumber(timeout, 'timeout', 0);
+  }
+
+  if (destination === threadId) {
+    throw new ERR_WORKER_SAME_THREAD();
+  }
+
+  const memory = new SharedArrayBuffer(WORKER_MESSAGING_SHARED_DATA);
+  const status = new Int32Array(memory);
+  const promise = AtomicsWaitAsync(status, WORKER_MESSAGING_STATUS_INDEX, 0, timeout).value;
+
+  const message = {
+    type: messageTypes.SEND_MESSAGE_TO_WORKER,
+    source: threadId,
+    destination,
+    value,
+    memory,
+    transferList,
+  };
+
+  if (isMainThread) {
+    handleMessageFromThread(message);
+  } else {
+    mainThreadPort.postMessage(message, transferList);
+  }
+
+  // Wait for the response
+  const response = await promise;
+
+  if (response === 'timed-out' || status[WORKER_MESSAGING_RESULT_INDEX] === 0) {
+    throw new ERR_WORKER_INVALID_ID(destination);
```

Comment:
not sure if it is worth/helpful to use different error codes for timeout, no handler, exception in handler.

Reply:
I reworked the error handling. We now have:

```
E('ERR_WORKER_MESSAGING_ERRORED', 'The destination thread threw an error while processing the message', Error);
E('ERR_WORKER_MESSAGING_FAILED', 'Cannot find the destination thread or listener', Error);
E('ERR_WORKER_MESSAGING_SAME_THREAD', 'Cannot sent a message to the same thread', Error);
E('ERR_WORKER_MESSAGING_TIMEOUT', 'Sending a message to another thread timed out', Error);
```